src/permissions/PermissionCallableAllowedContractNativeTokenRecurringAllowance.sol

```diff
@@ -89,20 +91,24 @@ contract PermissionCallableAllowedContractNativeTokenRecurringAllowance is
 
             if (selector == IPermissionCallable.permissionedCall.selector) {
                 // check call target is the allowed contract
-                if (call.target != values.allowedContract) revert UserOperationLib.TargetNotAllowed();
+                if (call.target != values.allowedContract) revert UserOperationLib.TargetNotAllowed(call.target);
                 // assume PermissionManager already prevents account as target
             } else if (selector == IMagicSpend.withdraw.selector) {
+                // check call target is MagicSpend
+                if (call.target != magicSpend) revert UserOperationLib.TargetNotAllowed(call.target);
+
                 // parse MagicSpend withdraw request
                 IMagicSpend.WithdrawRequest memory withdraw =
                     abi.decode(BytesLib.trimSelector(calls[i].data), (IMagicSpend.WithdrawRequest));
 
                 // check withdraw is native token
-                if (withdraw.asset != address(0)) revert InvalidWithdrawAsset();
+                if (withdraw.asset != address(0)) revert InvalidWithdrawAsset(withdraw.asset);
                 // do not need to accrue callsSpend because withdrawn value will be spent in other calls
             } else if (selector == IMagicSpend.withdrawGasExcess.selector) {
-                // ok
+                // check call target is MagicSpend
+                if (call.target != magicSpend) revert UserOperationLib.TargetNotAllowed(call.target);
```

Comment:
Key logic change: only should be able to call this selector on magic spend contract.